File: TypeScript/Arrays/enhancedRotateArray.ts
function reverse(array: number[], start: number, end: number): number[] {
    while (start < end) {
        [array[start], array[end]] = [array[end], array[start]];
        start++;
        end--;
    }
    return array;
}

function rotateArray(array: number[], k: number): number[] {
    if (array.length === 0) {
        return [];
    }

    k = k % array.length;

    reverse(array, 0, array.length - 1);
    reverse(array, 0, k - 1);
    reverse(array, k, array.length - 1);

    return array;
}

File: Golang/Arrays/enhancedRotateArray.go
package main

func reverse(array []int, start int, end int) {
	for start < end {
		array[start], array[end] = array[end], array[start]
		start++
		end--
	}
}

func enhancedRotateArray(array []int, k int) []int {
	if len(array) == 0 {
		return []int{}
	}

	k = k % len(array)
	if k == 0 {
		return array
	}

	reverse(array, 0, len(array)-1)
	reverse(array, 0, k-1)
	reverse(array, k, len(array)-1)

	return array
}
